specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+
+#### Signing workflow using plugin
+
+1. Given a user request to sign `oci-artifact`, with `key`
+1. Pull the image manifest using `oci-artifact` url, and construct a descriptor
+1. Append any user provided metadata and Notary metadata as descriptor annotations.
+1. Determine if the registered key uses a plugin
+1. Execute the plugin with `discover` command
+    1. If plugin supports capability `SIGNATURE_GENERATOR`
+        1. Generate the payload to be signed. For JWS this includes [JWS](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#supported-signature-envelopes) payload with `subject` claim as descriptor, additional JWS claims, and protected headers.
+        2. Execute the plugin with `generate-signature` command, set `request.key` to key definition, `request.payload` to the generated payload.
+        3. Validate the generated signature, return an error if any of the checks fails.
+           1. Check if `response.signingAlgorithm` is one of [supported signing algorithms](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection).
+           2. Check that the plugin did not modify `request.payload` before generating the signature, and the signature is valid for the given payload. Verify the hash of the `request.payload` against `response.signature`, using the public key of signing certificate (leaf certificate) in `response.certificateChain` along with the `response.signingAlgorithm`. This step does not include certificate chain validation (certificate chain leads to a trusted root configured in Notation), or revocation check.
+           3. Check that the `response.certificateChain` confirms to [Certificate Requirements].
+        4. Assemble the JWS Signature envelope using `response.signature`, `response.signingAlgorithm` and `response.certificateChain`. Notation may also generate and include timestamp signature in this step.
+        5. Generate a signature manifest for the given signature envelope.
+    2. Else if plugin supports capability `SIGNATURE_ENVELOPE_GENERATOR` *(covered in next section)*
+    3. Return an error
+
+***generate-signature***
+
+*Request*
+
+```json
+{
+  "contract-version" : <major-version.minor-version>,
+
+  // Complete key definition from
+  // /notation/config.json /signingKeys/keys with matching key name
+  "key" : <key definition>,
+
+  "payload" : <Base64 encoded payload to be signed>
+}
+```
+
+*key* : Required field that contains the complete key definition, which includes friendly name (`name`), key identifier (`id`) and any additional custom fields required by the plugin.
+
+*payload* : Required field that contains payload to be signed. This is currently the JWSPayload that includes the subject descriptor and other JWS claims.
+
+*Response*
+All response attributes are required.
+
+```json
+{    
+  // The same key id as passed in the request.
+  "keyId" : <key id>,
```

Comment:
Should we use `key` here instead of `keyId` as the key definition is passed as part of the request?

Reply:
The `key` in the request was actually the complete key definition, I'll rename that field to `keyDefinition` to be clearer.